src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -529,205 +538,171 @@ void checkConstraintLanguageVersionAndRuntimeSupportForConversion(SyntaxNode syn
-                    return BindListInterfaceCollectionExpression(node, targetType, wasCompilerGenerated: wasCompilerGenerated, elementType!, diagnostics);
-                case CollectionExpressionTypeKind.None:
-                    return BindCollectionExpressionForErrorRecovery(node, targetType, diagnostics);
-                default:
-                    throw ExceptionUtilities.UnexpectedValue(collectionTypeKind);
+                return BindCollectionExpressionForErrorRecovery(node, targetType, diagnostics);
             }
-        }
-
-        internal bool TryGetCollectionIterationType(ExpressionSyntax syntax, TypeSymbol collectionType, out TypeWithAnnotations iterationType)
-        {
-            BoundExpression collectionExpr = new BoundValuePlaceholder(syntax, collectionType);
-            return GetEnumeratorInfoAndInferCollectionElementType(
-                syntax,
-                syntax,
-                ref collectionExpr,
-                isAsync: false,
-                BindingDiagnosticBag.Discarded,
-                out iterationType,
-                builder: out _);
-        }
 
-        private BoundCollectionExpression BindArrayOrSpanCollectionExpression(
-            BoundUnconvertedCollectionExpression node,
-            TypeSymbol targetType,
-            bool wasCompilerGenerated,
-            CollectionExpressionTypeKind collectionTypeKind,
-            TypeSymbol elementType,
-            BindingDiagnosticBag diagnostics)
-        {
-            var syntax = node.Syntax;
+            var syntax = (ExpressionSyntax)node.Syntax;
+            MethodSymbol? collectionBuilderMethod = null;
 
             switch (collectionTypeKind)
             {
                 case CollectionExpressionTypeKind.Span:
                     _ = GetWellKnownTypeMember(WellKnownMember.System_Span_T__ctor_Array, diagnostics, syntax: syntax);
                     break;
+
                 case CollectionExpressionTypeKind.ReadOnlySpan:
                     _ = GetWellKnownTypeMember(WellKnownMember.System_ReadOnlySpan_T__ctor_Array, diagnostics, syntax: syntax);
                     break;
+
+                case CollectionExpressionTypeKind.CollectionBuilder:
+                    {
+                        Debug.Assert(elementType is { });
+
+                        var namedType = (NamedTypeSymbol)targetType;
+                        namedType.HasCollectionBuilderAttribute(out TypeSymbol? builderType, out string? methodName);
+                        var targetTypeOriginalDefinition = targetType.OriginalDefinition;
+                        TryGetCollectionIterationType(syntax, targetTypeOriginalDefinition, out TypeWithAnnotations elementTypeOriginalDefinition);
+                        var useSiteInfo = GetNewCompoundUseSiteInfo(diagnostics);
+                        collectionBuilderMethod = GetCollectionBuilderMethod(namedType, elementTypeOriginalDefinition.Type, builderType, methodName, ref useSiteInfo);
```

Comment:
>collectionBuilderMethod = GetCollectionBuilderMethod(namedType, elementTypeOriginalDefinition.Type, builderType, methodName, ref useSiteInfo); [](http://example.com/codeflow?start=24&length=142)

We should probably verify the collection builder method when calculating the conversion, and include the method in `CollectionExpressionUncommonData`, and report `ErrorCode.ERR_CollectionBuilderAttributeMethodNotFound` as a use-site error there.

Reply:
We can leave this as is, with the collection builder verified when binding the collection expression, since the other use-site error is now a binding error.